File: android/src/main/java/com/securitysuite/KeyAgreementProfile.java
package com.securitysuite;

import com.facebook.react.bridge.ReadableMap;

/**
 * Resolved key-agreement profile used for key storage and agreement.
 */
public final class KeyAgreementProfile {
  public enum Kind {
    P256_ECDH,
    X25519
  }

  public final Kind kind;
  public final String alias;
  public final String keyAgreementAlgorithm;
  public final String keyFactoryAlgorithm;

  private KeyAgreementProfile(
      Kind kind,
      String alias,
      String keyAgreementAlgorithm,
      String keyFactoryAlgorithm
  ) {
    this.kind = kind;
    this.alias = alias;
    this.keyAgreementAlgorithm = keyAgreementAlgorithm;
    this.keyFactoryAlgorithm = keyFactoryAlgorithm;
  }

  public static KeyAgreementProfile fromReadableMap(ReadableMap options) {
    if (options == null) {
      throw new IllegalArgumentException(
          "Crypto options are required. Call SecuritySuite.initialize() before crypto APIs."
      );
    }

    String keyAgreementAlgorithm = requireString(options, "keyAgreementAlgorithm");
    String keyFactoryAlgorithm = normalizeKeyFactoryAlgorithm(
        requireString(options, "keyFactoryAlgorithm")
    );

    switch (keyFactoryAlgorithm.toUpperCase() + "/" + keyAgreementAlgorithm.toUpperCase()) {
      case "EC/ECDH":
        return new KeyAgreementProfile(
            Kind.P256_ECDH,
            KeyAgreementKeyStore.legacyEcdhAlias(),
            keyAgreementAlgorithm,
            keyFactoryAlgorithm
        );
      case "OKP/X25519":
        return new KeyAgreementProfile(
            Kind.X25519,
            "com.securitysuite.x25519",
            keyAgreementAlgorithm,
            keyFactoryAlgorithm
        );
      default:
        throw new IllegalArgumentException(
            "Unsupported key profile: "
                + keyFactoryAlgorithm
                + "/"
                + keyAgreementAlgorithm
                + ". Supported profiles: EC/ECDH (P-256) and OKP/X25519."
        );
    }
  }

  public static KeyAgreementProfile fromCryptoConfig(CryptoConfig config) {
    com.facebook.react.bridge.WritableMap map = com.facebook.react.bridge.Arguments.createMap();
    map.putString("keyAgreementAlgorithm", config.keyAgreementAlgorithm);
    map.putString("keyFactoryAlgorithm", config.keyFactoryAlgorithm);
    return fromReadableMap(map);
  }

  private static String normalizeKeyFactoryAlgorithm(String value) {
    if ("Curve25519".equals(value) || "X25519".equals(value)) {
      return "OKP";
    }
    return value;
  }

  private static String requireString(ReadableMap map, String key) {
    if (!map.hasKey(key) || map.getString(key) == null) {
      throw new IllegalArgumentException("Missing required crypto option: " + key);
    }
    String value = map.getString(key).trim();
    if (value.isEmpty()) {
      throw new IllegalArgumentException("Missing required crypto option: " + key);
    }
    return value;
  }
}

File: ios/KeyAgreementProfile.swift
import Foundation
import CryptoKit

@available(iOS 13.0, *)
enum KeyAgreementProfileKind {
    case p256Ecdh
    case x25519
}

@available(iOS 13.0, *)
struct KeyAgreementProfile {
    let kind: KeyAgreementProfileKind
    let storageId: String
    let keyAgreementAlgorithm: String
    let keyFactoryAlgorithm: String

    static func from(dictionary: NSDictionary?) throws -> KeyAgreementProfile {
        guard let dictionary = dictionary else {
            throw NSError(domain: "KeyAgreementProfile", code: 0, userInfo: [
                NSLocalizedDescriptionKey:
                    "Crypto options are required. Call SecuritySuite.initialize() before crypto APIs.",
            ])
        }

        let keyAgreementAlgorithm = try requireString(dictionary, key: "keyAgreementAlgorithm")
        let keyFactoryAlgorithm = normalizeKeyFactoryAlgorithm(
            try requireString(dictionary, key: "keyFactoryAlgorithm")
        )

        switch (keyFactoryAlgorithm.uppercased(), keyAgreementAlgorithm.uppercased()) {
        case ("EC", "ECDH"):
            return KeyAgreementProfile(
                kind: .p256Ecdh,
                storageId: "ec-p256",
                keyAgreementAlgorithm: keyAgreementAlgorithm,
                keyFactoryAlgorithm: keyFactoryAlgorithm
            )
        case ("OKP", "X25519"):
            return KeyAgreementProfile(
                kind: .x25519,
                storageId: "okp-x25519",
                keyAgreementAlgorithm: keyAgreementAlgorithm,
                keyFactoryAlgorithm: keyFactoryAlgorithm
            )
        default:
            throw NSError(domain: "KeyAgreementProfile", code: 1, userInfo: [
                NSLocalizedDescriptionKey:
                    "Unsupported key profile: \(keyFactoryAlgorithm)/\(keyAgreementAlgorithm). "
                    + "Supported profiles: EC/ECDH (P-256) and OKP/X25519.",
            ])
        }
    }

    static func from(config: CryptoConfig) throws -> KeyAgreementProfile {
        switch (config.keyFactoryAlgorithm.uppercased(), config.keyAgreementAlgorithm.uppercased()) {
        case ("EC", "ECDH"):
            return KeyAgreementProfile(
                kind: .p256Ecdh,
                storageId: "ec-p256",
                keyAgreementAlgorithm: config.keyAgreementAlgorithm,
                keyFactoryAlgorithm: config.keyFactoryAlgorithm
            )
        case ("OKP", "X25519"):
            return KeyAgreementProfile(
                kind: .x25519,
                storageId: "okp-x25519",
                keyAgreementAlgorithm: config.keyAgreementAlgorithm,
                keyFactoryAlgorithm: config.keyFactoryAlgorithm
            )
        default:
            throw NSError(domain: "KeyAgreementProfile", code: 1, userInfo: [
                NSLocalizedDescriptionKey:
                    "Unsupported key profile: \(config.keyFactoryAlgorithm)/\(config.keyAgreementAlgorithm). "
                    + "Supported profiles: EC/ECDH (P-256) and OKP/X25519.",
            ])
        }
    }

    private static func normalizeKeyFactoryAlgorithm(_ value: String) -> String {
        switch value {
        case "Curve25519", "X25519":
            return "OKP"
        default:
            return value
        }
    }

    private static func requireString(_ dictionary: NSDictionary, key: String) throws -> String {
        guard let value = dictionary[key] as? String else {
            throw NSError(domain: "KeyAgreementProfile", code: 0, userInfo: [
                NSLocalizedDescriptionKey: "Missing required crypto option: \(key)",
            ])
        }
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            throw NSError(domain: "KeyAgreementProfile", code: 0, userInfo: [
                NSLocalizedDescriptionKey: "Missing required crypto option: \(key)",
            ])
        }
        return trimmed
    }
}

@available(iOS 13.0, *)
struct KeyAgreementKeyPair {
    let profile: KeyAgreementProfile
    let privateKeyData: Data
    let publicKeyData: Data

    func sharedSecret(with serverPublicKeyData: Data) throws -> SharedSecret {
        switch profile.kind {
        case .p256Ecdh:
            let privateKey = try P256.KeyAgreement.PrivateKey(derRepresentation: privateKeyData)
            let serverPublicKey = try P256.KeyAgreement.PublicKey(derRepresentation: serverPublicKeyData)
            return try privateKey.sharedSecretFromKeyAgreement(with: serverPublicKey)
        case .x25519:
            let privateKey = try Curve25519.KeyAgreement.PrivateKey(rawRepresentation: privateKeyData)
            let serverPublicKey = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: serverPublicKeyData)
            return try privateKey.sharedSecretFromKeyAgreement(with: serverPublicKey)
        }
    }
}
